metrics/multi_metrics.go

```diff
@@ -0,0 +1,76 @@
+package metrics
+
+// MultiMetrics is a metrics provider that sends metrics to at least one
+// underlying metrics provider (StoreMetrics). It can be configured to send
+// metrics to multiple providers at once.
+// It always starts with a StoreMetrics as the first provider, so that
+// Get and Store can depend on it.
+type MultiMetrics struct {
+	children []Metrics
+}
+
+func NewMultiMetrics() *MultiMetrics {
+	return &MultiMetrics{
+		children: []Metrics{&StoreMetrics{}},
+	}
+}
+
+// This is not safe for concurrent use!
+func (m *MultiMetrics) AddChild(met Metrics) {
+	m.children = append(m.children, met)
+}
+
+// This is not safe for concurrent use!
+func (m *MultiMetrics) Children() []Metrics {
+	return m.children
+}
+
+func (m *MultiMetrics) Register(name string, metricType string) {
+	for _, ch := range m.children {
+		ch.Register(name, metricType)
+	}
+}
+
+func (m *MultiMetrics) Increment(name string) { // for counters
+	for _, ch := range m.children {
+		ch.Increment(name)
+	}
+}
+
+func (m *MultiMetrics) Gauge(name string, val interface{}) { // for gauges
+	for _, ch := range m.children {
+		ch.Gauge(name, val)
+	}
+}
+
+func (m *MultiMetrics) Count(name string, n interface{}) { // for counters
+	for _, ch := range m.children {
+		ch.Count(name, n)
+	}
+}
+
+func (m *MultiMetrics) Histogram(name string, obs interface{}) { // for histogram
+	for _, ch := range m.children {
+		ch.Histogram(name, obs)
+	}
+}
+
+func (m *MultiMetrics) Up(name string) { // for updown
+	for _, ch := range m.children {
+		ch.Up(name)
+	}
+}
+
+func (m *MultiMetrics) Down(name string) { // for updown
+	for _, ch := range m.children {
+		ch.Down(name)
+	}
+}
+
+func (m *MultiMetrics) Get(name string) (float64, bool) { // for reading back a value
+	return m.children[0].Get(name) // this is the StoreMetrics
```

Comment:
Right now this is safe because there is no `RemoveChild` but it does feel a little risky to depend on this assumption for long into the future.  If StoreMetrics is always required, should it be a separate field on the struct and explicitly referenced in `Get`, `Store`, and each function before the loop?

```go
type MultiMetrics struct {
        store StoreMetrics
	children []Metrics
}

func (m *MultiMetrics) Histogram(name string, obs interface{}) { // for histogram
	m.store.Histogram(name, obs)
        for _, ch := range m.children {
		ch.Histogram(name, obs)
	}
}

func (m *MultiMetrics) Store(name string, val float64) { // for storing a rarely-changing value not sent as a metric
	m.Store(name, val)
}
```

Reply:
I originally had StoreMetrics as part of MultiMetrics, and then rewrote it to pull it out, because it felt like mixing concerns. This way, StoreMetrics implements Metrics, as does MultiMetrics, and the only special cases are the Get and Store implementations within MultiMetrics. I could have made Store do a loop, but that's pointless. I guess I could put a conditional in Get, but it's literally never going to fail.

I don't feel very strongly about this, so if you do, just say so and I'll change it.

Reply:
Fair enough. I've removed StoreMetrics and moved that logic into MultiMetrics. Had to delete a test because it depended on using StoreMetrics as a mock, but I don't think it won us much anyway.